### PR TITLE
feat(globalsearch): Implement tag-based filters in global search

### PR DESCRIPTION
**Pull request by tsdk02**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Added more explicit filters to global search:
Currently we rely on free form text search where opensearch matches the keywords across the entire document

Filters added currently:

- payment method
- currency
- status
- customer_email

Can map multiple options for every filter for global search

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Adding a more refined & strict search query where we do exact matches on fields provides better results

## How did you test it?

Tested using local data by sending cURL requests through Postman.
```bash
curl --location 'http://localhost:8080/analytics/v1/search' \
--header 'Accept: */*' \
--header 'Accept-Language: en-US,en;q=0.9' \
--header 'Connection: keep-alive' \
--header 'Content-Type: application/json' \
--header 'Origin: http://localhost:9000' \
--header 'Referer: http://localhost:9000/' \
--header 'Sec-Fetch-Dest: empty' \
--header 'Sec-Fetch-Mode: cors' \
--header 'Sec-Fetch-Site: same-site' \
--header 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36' \
--header 'api-key: hyperswitch' \
--header 'authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiZjQ1YWViMDgtOTBmNS00ZTM5LTljMjMtNDRkZmYxM2E4ODgwIiwibWVyY2hhbnRfaWQiOiJtZXJjaGFudF8xNzE4MDE4MjIyIiwicm9sZV9pZCI6Im9yZ19hZG1pbiIsImV4cCI6MTcxOTk5MTA0MSwib3JnX2lkIjoib3JnXzFuR09tR2N2dUxPVnBXUDJYMUFIIn0.VWP2Lj3US2qPb0nIZtpR2cHyfDpYBsR9lc9PvfQNdXY' \
--header 'sec-ch-ua: "Not/A)Brand";v="8", "Chromium";v="126", "Google Chrome";v="126"' \
--header 'sec-ch-ua-mobile: ?0' \
--header 'sec-ch-ua-platform: "macOS"' \
--data '{
    "query": "merchant_1718018222",
    "filters": {
        "currency": [
            "INR", "USD"
        ]
    }
}'
```
Screenshot 2024-06-28 at 1 48 53 AM
Screenshot 2024-06-28 at 1 51 11 AM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/analytics/src/search.rs [hunk 1]
```diff
@@ -82,9 +108,34 @@ pub async fn search_results(
...
+            query_builder
+                .add_filter_clause("status.keyword".to_string(), status)
+                .switch()?;
+        };
+        if let Some(payment_method) = filters.payment_method {
+            query_builder
+                .add_filter_clause("payment_method.keyword".to_string(), payment_method)
+                .switch()?;
+        };
+        if let Some(customer_email) = filters.customer_email {
```
Comment (lsampras) [inline crates/analytics/src/search.rs:133]:
```suggestion
        if let Some(customer_email) = filters.customer_email && !customer_email.is_empty() {
```
can we add this sort of validation to make sure that empty arrays are ignored?

**Review by lsampras — COMMENTED** [review]
rest looks good

File: crates/analytics/src/search.rs (hunk 1 above)
Comment (tsdk02) [reply crates/analytics/src/search.rs:147]: Yes, added the validations to ignore empty arrays

File: crates/analytics/src/search.rs [hunk 2]
```diff
@@ -20,9 +20,43 @@ pub async fn msearch_results(
...
+            if !currency.is_empty() {
+                query_builder
+                    .add_filter_clause("currency.keyword".to_string(), currency.clone())
+                    .switch()?;
+            }
+        };
+        if let Some(status) = filters.status {
+            if !status.is_empty() {
+                query_builder
+                    .add_filter_clause("status.keyword".to_string(), status.clone())
```
Comment (ivor-juspay) [inline crates/analytics/src/search.rs:40]: is the `.keyword` part required?

Comment (tsdk02) [reply crates/analytics/src/search.rs:40]: Yes, some of the fields require `.keyword`, and others don't, and we are unable to differentiate which fields require it. So adding `.keyword` for all the fields as of now.

**Review by lsampras — APPROVED** [review]

**Pull request merged**